src/app/utils/plug-descriptions.ts

```diff
@@ -0,0 +1,292 @@
+import { Perk } from 'app/clarity/descriptions/descriptionInterface';
+import { clarityDescriptionsSelector } from 'app/clarity/selectors';
+import { D2ManifestDefinitions } from 'app/destiny2/d2-definitions';
+import { settingSelector } from 'app/dim-api/selectors';
+import { useD2Definitions } from 'app/manifest/selectors';
+import { EXOTIC_CATALYST_TRAIT, modsWithConditionalStats } from 'app/search/d2-known-values';
+import { DestinyInventoryItemDefinition, ItemPerkVisibility } from 'bungie-api-ts/destiny2';
+import { ItemCategoryHashes, StatHashes } from 'data/d2/generated-enums';
+import perkToEnhanced from 'data/d2/trait-to-enhanced-trait.json';
+import _ from 'lodash';
+import { useSelector } from 'react-redux';
+
+interface DimPlugPerkDescription {
+  perkHash: number;
+  name?: string;
+  description?: string;
+  requirement?: string;
+}
+
+interface DimPlugDescriptions {
+  perks: DimPlugPerkDescription[];
+  communityInsight: Perk | undefined;
+}
+
+// some stats are often referred to using different names
+const statNameAliases = {
+  [StatHashes.AimAssistance]: ['Aim Assist'],
+  [StatHashes.AmmoCapacity]: ['Magazine Stat'],
+  [StatHashes.ReloadSpeed]: ['Reload'],
+};
+
+export const enhancedPerkToRegularPerk = _.mapValues(_.invert(perkToEnhanced), Number);
+
+export function usePlugDescriptions(
+  plug?: DestinyInventoryItemDefinition,
+  stats?: {
+    value: number;
+    statHash: number;
+  }[]
+): DimPlugDescriptions {
+  const defs = useD2Definitions();
+  const allClarityDescriptions = useSelector(clarityDescriptionsSelector);
+  const descriptionsToDisplay = useSelector(settingSelector('descriptionsToDisplay'));
+
+  const result: DimPlugDescriptions = {
+    perks: [],
+    communityInsight: undefined,
+  };
+
+  if (!plug || !defs) {
+    return result;
+  }
+
+  const showBungieDescription =
+    !$featureFlags.clarityDescriptions || descriptionsToDisplay !== 'community';
+  const showCommunityDescription =
+    $featureFlags.clarityDescriptions && descriptionsToDisplay !== 'bungie';
+  const showCommunityDescriptionOnly =
+    $featureFlags.clarityDescriptions && descriptionsToDisplay === 'community';
+
+  // within this plug, let's not repeat any strings
+  const statStrings = new Set<string>();
+
+  if (stats) {
+    // preload the used string tracker with common text representations of stat modifications
+    for (const stat of stats) {
+      const statDef = defs.Stat.get(stat.statHash);
+      if (statDef) {
+        const statNames = [statDef.displayProperties.name].concat(statNameAliases[stat.statHash]);
+        for (const statName of statNames) {
+          if (stat.value < 0) {
+            statStrings.add(`${stat.value} ${statName}`);
+            statStrings.add(`${stat.value} ${statName} ▼`);
+          } else {
+            statStrings.add(`+${stat.value} ${statName}`);
+            statStrings.add(`+${stat.value} ${statName} ▲`);
+            statStrings.add(`Grants ${stat.value} ${statName}`);
```

Comment:
Will this only work for English strings? Or is this meant to only match Clarity descriptions?

Reply:
this is stripping stuff from clarity desc that's already covered by
- bungie descriptions (if they're also displayed)
- our stat displays (always)

it only really works in english, afaik clarity stuff is only in english?

Reply:
Unless someone wants to translate its only English

Reply:
tbh I don't recommend doing that unless you want to fix stuff over and over again

Reply:
lest take "compulsive reloader" as an example
if you take 50 reload speed from this and display it without context it will look like it's passive +50 reload
not only that you skip 0.95 multiplayer which is just as important 
multiplayer is final reload time * multiplayer // 2s * 0.95 = 1.9s not allot in this case but it works even if you have 100 stat

While above 70% Magazine: 
50 Reload Speed with a 0.95x Duration Scale 

Enhanced: Reload increased to [55 maybe?]

Reply:
> if you take 50 reload speed from this and display it without context it will look like it's passive +50 reload

Just to clarify, we are not extracting stat strings out of community descriptions and pretending they are part of the plug's investment stats. This code is identifying where a community description (or Bungie description) contains text that matches an investment stat modification **that already exists on the plug** and strips it out so we don't show the same information twice.

For example, Icarus Grip's community description is `Grants 15 Airborne Effectiveness` but it already has an investment stat entry for `+15 Airborne Effectiveness`. In this case, we would hide the community description because it would be repeating the same information.

You mentioned earlier that:

> Perks and or mods with investment stats (example 15 aim assist) will get special marker for easy removal

This sounds great and I'm keen to use these markers when they're available instead of trying to match strings in the community description. That being said, we will probably still need to keep this code to handle stat strings in Bungie descriptions (e.g. `+10 Mobility ▲` in `Mobility Mod` and `+10 Range` in `Adept Range` mod)